### ISSUE TITLE
Pythran makes np.searchsorted much slower?

### ISSUE THREAD

**Issue by charlotte12l** [serge-sans-paille/pythran#1793]
I tried to use Pythran to improve a function called `_cdf_distance`. When the input size is small (e.g. 4), Pythran is 3x faster than the original python code. However, when the size is large (e.g. 4000), Pythran is 3x slower than the original code. :(

I'm still trying to figure out what causes such a situation, and here is my timing report and codes. https://github.com/charlotte12l/pythran_tests/blob/main/_cdf_distance/timing.ipynb

**Comment by rgommers** [comment]
Looking at the implementation at https://github.com/scipy/scipy/pull/14154, there's no for-loops, so it's really about how the (vectorized) NumPy and Pythran function calls and indexing compare. I would expect Pythran to at least not be slower for large arrays.

Line profiling this Pythran code is something I'm not sure is possible, but just taking the kernel and commenting more lines out should be relatively straightforward here. E.g benchmark just 1-2 lines at a time. First:
```python
u_sorter = np.argsort(u_values)
v_sorter = np.argsort(v_values)
```
then:
```python
u_sorter = np.argsort(u_values)
v_sorter = np.argsort(v_values)

all_values = np.concatenate((u_values, v_values))
```
etc.

Maybe it's also worth inspecting the optimized Python code: https://pythran.readthedocs.io/en/latest/MANUAL.html#read-the-optimized-python-code

**Comment by charlotte12l** [comment]
Thank you so much!!!  I tried the first method and I think `if` statement is the cause of the problem.

When I benchmark the following lines with size=4000, pythran vs original is 8.74 µs vs 1.35ms.(`%timeit -n 100 -r 3 _pythran_cdf_distance(1,u_values, v_values)`)

```
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)

    all_values = np.concatenate((u_values, v_values))
    all_values = np.sort(all_values, kind='mergesort')

    # Compute the differences between pairs of successive values of u and v.
    deltas = np.diff(all_values)

    # Get the respective positions of the values of u and v among the values of
    # both distributions.
    u_cdf_indices = np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')
    v_cdf_indices = np.searchsorted(v_values[v_sorter], all_values[:-1], 'right')

    # Calculate the CDFs of u and v using their weights, if specified.
#     if u_weights is None:
    u_cdf = u_cdf_indices / u_values.size
```

 If I uncommented `if u_weights is None`,  pythran vs original is 2.02 ms vs 1.19 ms.(`%timeit -n 100 -r 3 _pythran_cdf_distance(1,u_values, v_values)`)

```
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)

    all_values = np.concatenate((u_values, v_values))
    all_values = np.sort(all_values, kind='mergesort')

    # Compute the differences between pairs of successive values of u and v.
    deltas = np.diff(all_values)

    # Get the respective positions of the values of u and v among the values of
    # both distributions.
    u_cdf_indices = np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')
    v_cdf_indices = np.searchsorted(v_values[v_sorter], all_values[:-1], 'right')

    # Calculate the CDFs of u and v using their weights, if specified.
    if u_weights is None:
        u_cdf = u_cdf_indices / u_values.size

```

For the second method, I will get some strange errors if I use `pythran -P arc_distance.py | yapf`. I'm still trying to fix that.

```
(scipydev) charlotte@CHARLOTLIU-MB0 _cdf_distance % pythran -P _pythran_cdf_distance.py | yapf
Traceback (most recent call last):
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 115, in ParseCodeToTree
    tree = parser_driver.parse_string(code, debug=False)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/driver.py", line 103, in parse_string
    return self.parse_tokens(tokens, debug)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/driver.py", line 71, in parse_tokens
    if p.addtoken(type, value, (prefix, start)):
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/parse.py", line 119, in addtoken
    ilabel = self.classify(type, value, context)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/parse.py", line 175, in classify
    raise ParseError("bad token", type, value, context)
lib2to3.pgen2.parse.ParseError: bad token: type=58, value=' ', context=('', (7, 78))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 121, in ParseCodeToTree
    tree = parser_driver.parse_string(code, debug=False)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/driver.py", line 103, in parse_string
    return self.parse_tokens(tokens, debug)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/driver.py", line 71, in parse_tokens
    if p.addtoken(type, value, (prefix, start)):
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/parse.py", line 119, in addtoken
    ilabel = self.classify(type, value, context)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/lib2to3/pgen2/parse.py", line 175, in classify
    raise ParseError("bad token", type, value, context)
lib2to3.pgen2.parse.ParseError: bad token: type=58, value=' ', context=('', (7, 78))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/charlotte/anaconda3/envs/scipydev/bin/yapf", line 10, in <module>
    sys.exit(run_main())
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/__init__.py", line 362, in run_main
    sys.exit(main(sys.argv))
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/__init__.py", line 104, in main
    reformatted_source, _ = yapf_api.FormatCode(
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/yapflib/yapf_api.py", line 147, in FormatCode
    tree = pytree_utils.ParseCodeToTree(unformatted_source)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 127, in ParseCodeToTree
    raise e
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/site-packages/yapf/yapflib/pytree_utils.py", line 125, in ParseCodeToTree
    ast.parse(code)
  File "/Users/charlotte/anaconda3/envs/scipydev/lib/python3.8/ast.py", line 47, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 2
    def _pythran_cdf_distance(p, u_values, v_values, u_weights=builtins.None, v_weights=builtins.None):
                                                                        ^
SyntaxError: invalid syntax
```

**Comment by rgommers** [comment]
> When I benchmark the following lines with input size=4000, pythran vs original is 8.74 µs vs 1.35ms.(`%timeit -n 100 -r 3 _pythran_cdf_distance(1,u_values, v_values)`)

I don't think that's possible - probably Pythran is not doing anything because you didn't include a `return` statement (it uses lazy evaluation, so no result means no calculations needed). I just tried, it takes about ~1.5 ms if you keep the `return np.sum ...` statement.

**Comment by rgommers** [comment]
```
    def _pythran_cdf_distance(p, u_values, v_values, u_weights=builtins.None, v_weights=builtins.None):
                                                                        ^
SyntaxError: invalid syntax
```
is a bug in the generated Python code. Using `builtins.None` was valid in Python 2, but not in Python 3. It still exists there, but you cannot access it via the namespace:
```python
>>> import builtins
>>> none = getattr(builtins, "None")
>>> none is None
True
>>> builtins.None
  File "<ipython-input-11-ad745c64d6f0>", line 1
    builtins.None
             ^
SyntaxError: invalid syntax
```

**Comment by charlotte12l** [comment]
> I don't think that's possible - probably Pythran is not doing anything because you didn't include a `return` statement (it uses lazy evaluation, so no result means no calculations needed). I just tried, it takes about ~1.5 ms if you keep the `return np.sum ...` statement.

Sorry for my mistake. Yes, you are right! After trying line by line, I think `np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')` is the main cause.

**Comment by rgommers** [comment]
I'm still fairly new to optimizing Pythran code, so I'm not sure how useful it is to run `pythran -P`. But it's interesting at least:)

I tried `pythran -P _cdf_distance_1d.py > tmp.py` and then edited it lightly (e.g., replacing `__pythran_import_numpy` with `np`) to make it readable. This is what comes out:
```python
import numpy as np


def _cdf_distance_1d(p, u_values, v_values, u_weights=None, v_weights=None):
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)
    all_values_ = np.sort(np.concatenate((u_values, v_values)), (-1), 'mergesort')
    deltas = np.diff(all_values_)
    __tuple0 = if(is_none(u_weights), $isstatic0, $isstatic1)(np.searchsorted(u_values[u_sorter], all_values_[:(-1)], 'right'), u_sorter, u_values, u_weights)
    u_cdf = __tuple0[0]

    __tuple1 = if(is_none(v_weights), $isstatic2, $isstatic3)(np.searchsorted(v_values[v_sorter], all_values_[:(-1)], 'right'), v_sorter, v_values, v_weights)
    v_cdf = __tuple1[0]
    if (p == 1):
        return np.sum(np.multiply(np.abs((u_cdf - v_cdf)), deltas))
    else:
        if (p == 2):
            return np.sqrt(np.sum(np.multiply(np.square((u_cdf - v_cdf)), deltas)))
        return np.power(np.sum(np.multiply(np.power(np.abs((u_cdf - v_cdf)), p), deltas)), (1 / p))


def $isstatic0(u_cdf_indices, u_sorter, u_values, u_weights):
    return ((u_cdf_indices / builtins.getattr(u_values, 'size')),)

def $isstatic1(u_cdf_indices, u_sorter, u_values, u_weights):
    u_sorted_cumweights = np.concatenate(([0], np.cumsum(u_weights[u_sorter])))
    return ((u_sorted_cumweights[u_cdf_indices] / u_sorted_cumweights[(-1)]),)

def $isstatic2(v_cdf_indices, v_sorter, v_values, v_weights):
    return ((v_cdf_indices / builtins.getattr(v_values, 'size')),)

def $isstatic3(v_cdf_indices, v_sorter, v_values, v_weights):
    v_sorted_cumweights = np.concatenate(([0], np.cumsum(v_weights[v_sorter])))
    return ((v_sorted_cumweights[v_cdf_indices] / v_sorted_cumweights[(-1)]),)
```

**Comment by rgommers** [comment]
> After trying line by line, I think `np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')` is the main cause. pythran vs original is 3.01 ms vs 1.17 ms

That does sound plausible. So either `searchsorted` implementation is a bit slow, or the advanced indexing operation (`u_values[u_sorter]`) is.

If you take the `np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')` line and split it into multiple lines, you can figure out which operation is the problem here.

**Comment by charlotte12l** [comment]
> 
> That does sound plausible. So either `searchsorted` implementation is a bit slow, or the advanced indexing operation (`u_values[u_sorter]`) is.
> 
> If you take the `np.searchsorted(u_values[u_sorter], all_values[:-1], 'right')` line and split it into multiple lines, you can figure out which operation is the problem here.

Thanks for your advice! I split them into multiple lines and I find Pythran did make np.searchsort much slower.

For the following code, Pythran vs original is 1.07 ms vs 767 µs.
```
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)
    
    all_values = np.concatenate((u_values, v_values))
    all_values = np.sort(all_values, kind='mergesort')

    # Compute the differences between pairs of successive values of u and v.
    deltas = np.diff(all_values)
    
    tmp_u = u_values[u_sorter]
    values_tmp = all_values[:-1]
    return tmp_u, values_tmp, deltas
```

However, for the following code, Pythran vs original is 2.06 ms vs 987 µs.

```
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)
    

    all_values = np.concatenate((u_values, v_values))
    all_values = np.sort(all_values, kind='mergesort')

    # Compute the differences between pairs of successive values of u and v.
    deltas = np.diff(all_values)
    
    tmp_u = u_values[u_sorter]
    values_tmp = all_values[:-1]
    u_cdf_indices = np.searchsorted(tmp_u, values_tmp, 'right')
    
    return u_cdf_indices, deltas

```

**Comment by charlotte12l** [comment]
I tested a small example:

```
%pythran_searchsort.py
import numpy as np

#pythran export pythran_searchsort(float[:], float[:])
def pythran_searchsort(a1, a2):
    return np.searchsorted(a1, a2, side='right')
```

The timing shows Pythran version is slower than the original:

```
from pythran_searchsort import pythran_searchsort
import numpy as np

size = 4000
rng = np.random.default_rng(12345678)
a = rng.random(size) * 10
b = rng.random(size) * 10
```
```
%timeit -n 100 -r 3 pythran_searchsort(a,b)
674 µs ± 43.2 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```
```
%timeit -n 100 -r 3 np.searchsorted(a,b,side='right')
280 µs ± 27.7 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```

**Comment by rgommers** [comment]
It looks like the numpy implementation (`PyArray_SearchSorted`) uses binary search, while Pythran's `searchsorted` does a linear scan.

**Comment by serge-sans-paille** [comment]
Mmmmh, pythran is using `lower_bound` so that's a binary search too. Anyway, @charlotte12l provided a very nice reproducer, I should be able to investigate.

**Comment by serge-sans-paille** [comment]
Should be fixed by #1794 . @charlotte12l can you confirm?
Btw in you benhcmark, you should sort `a` otherwise the assumption of `serachsorted` are not met.
Thanks **a lot** for the investigation and reproducer.

**Comment by charlotte12l** [comment]
@serge-sans-paille Thanks, it works!
The timing shows Pythran version is faster than the original on the new benchmark now!
```
from pythran_searchsort import pythran_searchsort
import numpy as np

size = 4000
rng = np.random.default_rng(12345678)
a = rng.random(size) * 10
b = rng.random(size) * 10
a = np.sort(a) # sort a
```

```
%timeit -n 100 -r 3 pythran_searchsort(a,b)
265 µs ± 6.77 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```
```
%timeit -n 100 -r 3 np.searchsorted(a,b,side='right')
385 µs ± 11 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```

However for these following lines in `_cdf_distance`, pythran is still slower on my local machine, and I think it is mainly due to the indexing operations(If I `return tmp_u, values_tmp, deltas` before `np.searchsorted`,  pythran is already slower than original- 1.07 ms vs 767 µs). 

For the following lines, before your fix, pythran vs original is 2.06ms vs 987µs, now is 1.31ms vs 955 µs.

```
    u_sorter = np.argsort(u_values)
    v_sorter = np.argsort(v_values)
    
    all_values = np.concatenate((u_values, v_values))
    all_values = np.sort(all_values, kind='mergesort')

    # Compute the differences between pairs of successive values of u and v.
    deltas = np.diff(all_values)
    
    tmp_u = u_values[u_sorter]
    values_tmp = all_values[:-1]
#     return tmp_u, values_tmp, deltas
    
    u_cdf_indices = np.searchsorted(tmp_u, values_tmp, 'right')
    
    return u_cdf_indices, deltas
```

**Comment by serge-sans-paille** [comment]
Great! The indexing operation is indeed a pain, I'll dig into that one!
